### PR TITLE
A/B shape skip based on SQ/H/V shape costs

### PR DESCRIPTION
**Pull request by JackZhouVCD**
## Description
Determine if the evaluation of a and b shapes could be skipped based on cost of square, h and v shapes. Namely:
 - skip HA and HB if h_cost > (weighted sq_cost)
 - skip VA and VB if v_cost > (weighted sq_cost)

Weighting to sq cost can be controlled with the `-sqw` flag, which is expressed as a percentage and set to `100` (neutral) by default.
 - To turn off this feature, set `-sqw 0`
 - To use *aggressive* weighting, set `-sqw` to a value lesser than 100
     - faster but more BD-Rate loss
 - To use *conservative* weighting, set `-sqw` to a value greater than 100
     - slower but less BD-Rate loss

## Author

@hguermaz 
@JackZhouVCD 

## Type of change

Enhancement

## Tests and performance
All data generated from full-list objective-1-fast dataset in the context of enc-mode 0. Speed data is generated using AWS ec2 48 vCPU instances

| | Deviation|
|-|-|
| **Speed** | 13.58% |
| **BD-Rate** | 0.106% |

### CODE REVIEW

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 1]
```diff
@@ -33,6 +33,8 @@
 extern "C" {
 #endif
 
+#define LESS_RECTANGULAR_CHECK_LEVEL 1
```
Comment (hassount) [inline Source/Lib/Common/Codec/EbDefinitions.h]: can we get a description for this macro?

File: Source/Lib/Common/Codec/EbEncDecProcess.c [hunk 2]
```diff
@@ -1535,6 +1535,22 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(
...
 #endif
 
+
+#if LESS_RECTANGULAR_CHECK_LEVEL
+
+    // Weighting (expressed as a percentage) applied to 
+    // square shape costs for determining if a and b 
+    // shapes should be skipped. Namely:
+    // skip HA and HB if h_cost > (weighted sq_cost)
+    // skip VA and VB if v_cost > (weighted sq_cost)
```
Comment (hassount) [inline Source/Lib/Common/Codec/EbEncDecProcess.c]: can we add a config flag for this feature? -1: auto - 0: off - N: threshold

Comment (JackZhouVCD) [reply Source/Lib/Common/Codec/EbEncDecProcess.c]: done, added the `-sqw` flag

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 1 above)
Comment (JackZhouVCD) [reply Source/Lib/Common/Codec/EbDefinitions.h]: done. See latest commit

**Comment by EwoutH** [comment]
@JackZhouVCD Very nice work. Could you make sure that there are no trailing white spaces in your commit messages though?

**Comment by JackZhouVCD** [comment]
> @JackZhouVCD Very nice work. Could you make sure that in the future there are no trailing white spaces in your commit messages though?

sure, I'll make sure of that. However, I checked my commit messages and I'm not seeing any trailing spaces. Can you point me towards where I have trailing spaces? I can still reword the commit messages before this gets merged.

**Comment by EwoutH** [comment]
Screenshot_408
Here's one, and git says it found a few more:

```
ewoutterhoeven@av1-bench-cas-48:~/SVT-AV1$ git apply --reject --whitespace=fix 762.patch
762.patch:35: trailing whitespace.
 * cumulatively updates rd-cost of each enumerated 
762.patch:36: trailing whitespace.
 * partitioning to decide if nsq should be skipped 
762.patch:136: trailing whitespace.
    // Weighting (expressed as a percentage) applied to 
762.patch:137: trailing whitespace.
    // square shape costs for determining if a and b 
762.patch:141: trailing whitespace.
```

**Comment by JackZhouVCD** [comment]
> Screenshot_408

That's strange, when I reword my commit messages in my editor it doesn't show any trailing spaces. Regardless, I did a bit of modifications and force pushed my branch. Hopefully it's fine now. I believe that the space that can be highlighted at the end of the commit you're showing is for a newline character since I edited my message and made sure there were no trailing spaces (and it's still showing).

> Here's one, and git says it found a few more:
> ```
> ewoutterhoeven@av1-bench-cas-48:~/SVT-AV1$ git apply --reject --whitespace=fix 762.patch
> 762.patch:35: trailing whitespace.
>  * cumulatively updates rd-cost of each enumerated 
> 762.patch:36: trailing whitespace.
>  * partitioning to decide if nsq should be skipped 
> 762.patch:136: trailing whitespace.
>     // Weighting (expressed as a percentage) applied to 
> 762.patch:137: trailing whitespace.
>     // square shape costs for determining if a and b 
> 762.patch:141: trailing whitespace.
> ```

As for this, can you get a more updated patch and check again? those comments no longer exist in the code, as you can see [here](https://github.com/OpenVisualCloud/SVT-AV1/pull/762/files)

**Comment by EwoutH** [comment]
I ran a [quick benchmark](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=2005362768). EncModes 8, 6 and 4 show no significant improvement, but 2 and 0 show between 5,6% and 24,3% speedup from -sqw 100 and lower.

| Normalized | -enc-mode 8 | -enc-mode 6 | -enc-mode 4 | -enc-mode 2 | -enc-mode 0 |
|------------|------------:|------------:|------------:|------------:|------------:|
| -sqw 0     |       1,000 |       1,000 |       1,000 |       1,000 |       1,000 |
| -sqw 400   |       0,983 |       1,011 |       1,006 |       1,007 |       0,993 |
| -sqw 200   |       0,986 |       1,022 |       1,021 |       0,997 |       0,995 |
| -sqw 100   |       0,993 |       1,009 |       0,992 |       1,056 |       1,135 |
| -sqw 50    |       0,986 |       1,007 |       0,999 |       1,087 |       1,223 |
| -sqw 25    |       0,995 |       0,996 |       1,001 |       1,088 |       1,243 |

**Comment by EwoutH** [comment]
> As for this, can you get a more updated patch and check again? those comments no longer exist in the code, as you can see [here](https://github.com/OpenVisualCloud/SVT-AV1/pull/762/files)

No change when I try to apply the new patch unfortunately.
```
ewoutterhoeven@av1-bench-cas-16-enc-2:~/SVT-AV1$ git apply --reject --whitespace=fix 762.patch
762.patch:35: trailing whitespace.
 * cumulatively updates rd-cost of each enumerated 
762.patch:36: trailing whitespace.
 * partitioning to decide if nsq should be skipped 
762.patch:136: trailing whitespace.
    // Weighting (expressed as a percentage) applied to 
762.patch:137: trailing whitespace.
    // square shape costs for determining if a and b 
762.patch:141: trailing whitespace.
```

**Comment by JackZhouVCD** [comment]
> > As for this, can you get a more updated patch and check again? those comments no longer exist in the code, as you can see [here](https://github.com/OpenVisualCloud/SVT-AV1/pull/762/files)
> 
> No change when I try to apply the new patch unfortunately.
> 
> ```
> ewoutterhoeven@av1-bench-cas-16-enc-2:~/SVT-AV1$ git apply --reject --whitespace=fix 762.patch
> 762.patch:35: trailing whitespace.
>  * cumulatively updates rd-cost of each enumerated 
> 762.patch:36: trailing whitespace.
>  * partitioning to decide if nsq should be skipped 
> 762.patch:136: trailing whitespace.
>     // Weighting (expressed as a percentage) applied to 
> 762.patch:137: trailing whitespace.
>     // square shape costs for determining if a and b 
> 762.patch:141: trailing whitespace.
> ```

I see, I guess this is caused by one of the earlier commits in the PR, what ends up being applied to master won't contain these. Or, to make things cleaner, we can look into squashing the PR when merging

**Comment by hguermaz** [comment]
@EwoutH Thank you for the additional speed data !

Yes we expect higher gains for M0 where an exhaustive NSQ is used. This speed optimization should not impact EncModes 3 and beyond as NSQ is currently OFF for ncModes 3 and beyond.

**Pull request merged**